### PR TITLE
Split vertical fluxes on basin

### PR DESCRIPTION
**Pull request by SouthEndMusic**
Fixes https://github.com/Deltares/Ribasim/issues/661.

### CODE REVIEW

**Comment by SouthEndMusic** [comment]
@visr let me know whether you think this is going in the right direction. The vertical fluxes get integrated now but they are not yet saved. Something about this also feels a bit silly, since this does an approximation of the integral while we could easily obtain the exact value.

**Comment by SouthEndMusic** [comment]
In fact, for forward fill we don't need `vertical_flux_prev` at all, we just have to make sure the integration is computed before `vertical_flux` is updated.

**Comment by SouthEndMusic** [comment]
Maybe https://github.com/Deltares/Ribasim/issues/1084 can also be picked up in this PR, just expose `vertical_flux_integrated` in stead of `vertical_flux` values. One thing we need to think about as discussed with @HendrikKok is what happens when a basin dries out.

File: core/src/callback.jl [hunk 1]
```diff
@@ -496,6 +491,12 @@ function update_basin(integrator)::Nothing
         set_table_row!(table, row, i)
     end
 
+    for (i, id) in enumerate(basin.node_id)
+        update_vertical_flux!(basin, storage, i)
+    end
+
+    # Forget about vertical fluxes before basin update
+    copyto!(vertical_flux_prev, vertical_flux)
```
Comment (SouthEndMusic) [inline core/src/callback.jl:499]: To be most accurate, maybe an exception should be made for vertical fluxes which are not set by the `Basin \ time` table.

File: core/src/bmi.jl [hunk 2]
```diff
@@ -36,10 +36,14 @@ function BMI.get_value_ptr(model::Model, name::AbstractString)
...
     elseif name == "basin.level"
         get_tmp(model.integrator.p.basin.current_level, 0)
-    elseif name == "basin.infiltration"
-        model.integrator.p.basin.infiltration
-    elseif name == "basin.drainage"
-        model.integrator.p.basin.drainage
+    elseif name == "basin.infiltration_instantaneous"
+        get_tmp(model.integrator.p.basin.vertical_flux, 0).infiltration
+    elseif name == "basin.drainage_instantaneous"
+        get_tmp(model.integrator.p.basin.vertical_flux, 0).drainage
```
Comment (visr) [inline core/src/bmi.jl:42]:
```suggestion
    elseif name == "basin.infiltration"
        get_tmp(model.integrator.p.basin.vertical_flux, 0).infiltration
    elseif name == "basin.drainage"
        get_tmp(model.integrator.p.basin.vertical_flux, 0).drainage
```

File: core/test/bmi_test.jl [hunk 3]
```diff
@@ -61,8 +61,10 @@ end
     for name in [
         "basin.storage",
         "basin.level",
-        "basin.infiltration",
-        "basin.drainage",
+        "basin.infiltration_instantaneous",
+        "basin.drainage_instantaneous",
```
Comment (visr) [inline core/test/bmi_test.jl:65]:
```suggestion
        "basin.infiltration",
        "basin.drainage",
```

File: core/src/parameter.jl [hunk 4]
```diff
@@ -171,26 +173,28 @@ struct Basin{T, C} <: AbstractParameterNode
...
         node_id,
-        precipitation,
-        potential_evaporation,
-        drainage,
-        infiltration,
+        vertical_flux_from_input::V1,
+        vertical_flux::V2,
+        vertical_flux_prev::V3,
+        vertical_flux_integrated,
+        vertical_flux_bmi,
```
Comment (visr) [inline core/src/parameter.jl:180]:
```suggestion
        vertical_flux_integrated::V3,
        vertical_flux_bmi::V3,
```

File: core/src/solve.jl [hunk 5]
```diff
@@ -52,34 +51,43 @@ end
...
+    return influx
+end
+
+function formulate_basins!(
+    du::AbstractVector,
+    basin::Basin,
+    storage::AbstractVector,
+)::Nothing
+    for (i, id) in enumerate(basin.node_id)
+        # add all precipitation that falls within the profile
```
Comment (visr) [inline core/src/solve.jl:88]:
```suggestion
    for i in eachindex(basin.node_id)
```

File: core/src/callback.jl [hunk 6]
```diff
@@ -496,6 +493,12 @@ function update_basin(integrator)::Nothing
         set_table_row!(table, row, i)
     end
 
+    for (i, id) in enumerate(basin.node_id)
+        update_vertical_flux!(basin, storage, i)
+    end
```
Comment (visr) [inline core/src/callback.jl:498]:
```suggestion
    for i in eachindex(basin.node_id)
        update_vertical_flux!(basin, storage, i)
    end
```

File: core/src/callback.jl [hunk 7]
```diff
@@ -496,6 +493,12 @@ function update_basin(integrator)::Nothing
...
+
+    # Forget about vertical fluxes before basin update
```
Comment (visr) [inline core/src/callback.jl:500]:
```suggestion
    # Forget about vertical fluxes to handle discontinuous forcing from basin_update
```

File: core/src/callback.jl (hunk 1 above)
Comment (visr) [reply core/src/callback.jl:501]: Typically only some of the vertical fluxes will be in `Basin \ time`. Could you make an issue about this? I don't fully oversee the error introduced in this case.

File: core/src/util.jl [hunk 8]
```diff
@@ -707,3 +707,25 @@ function Base.get(
         nothing
     end
 end
+
+"""
+Get the time interval between (flow) saves
+"""
+function get_Δt(integrator, graph)::Float64
```
Comment (visr) [inline core/src/util.jl:714]: Probably better to make this accept only `integrator` and take the graph out of the parameters.

File: core/src/allocation_optim.jl [hunk 9]
```diff
@@ -321,9 +321,15 @@ function get_basin_data(
...
 )
     (; graph, basin, level_demand) = p
+    (; vertical_flux) = basin
     (; Δt_allocation) = allocation_model
     @assert node_id.type == NodeType.Basin
-    influx = get_flow(graph, node_id, 0.0)
+    vertical_flux = get_tmp(vertical_flux, 0)
+    _, basin_idx = id_index(basin.node_id, node_id)
+    # NOTE: Instantaneous
+    influx =
```
Comment (visr) [inline core/src/allocation_optim.jl:330]: Is it possible to pull out this as a `get_flow` method and use it in `update_vertical_flux!` as well?

**Review by visr — APPROVED** [review]

File: docs/core/usage.qmd [hunk 10]
```diff
@@ -705,21 +705,30 @@ derivative       | Float64  | -        | -
 
 ## Basin - `basin.arrow`
 
-The basin table contains results of the storage and level of each basin at every solver
-timestep. The initial condition is also written to the file.
+The basin table contains:
+
+- results of the storage and level of each basin, which are instantaneous values;
+- results of the vertical fluxes on each basin, which are mean values over the `saveat` intervals. In the time column the start of the period is indicated. This means that for the final time in the table the mean vertical fluxes are not computed, and thus left $0$.
```
[the remark this answers is not in the document]
Comment (SouthEndMusic) [reply docs/core/usage.qmd:711]: Ah missing makes more sense, this is just how it worked out in my implementation.

**Pull request merged**